### ISSUE TITLE
Please enable use of CONSTANTS in call to size

### ISSUE THREAD

**Issue by jarmohurri** [processing/processing#3632]
Greetings.

I just installed 3.0 beta 3 and noticed that the use of constants is no longer supported in size(). I then read that variables are no longer supported as parameters to size().

Could you please, *please* enable the use of constants in call to size(). As a programming teacher I have several reasons for this. One would be the following example:

    final int H = 100;
    size (2 * H, H);

This is equivalent to `size (200, 100);`, but it is much better: the use of constants enables me to retain the 2:1 aspect ratio when changing the size of the sketch, and changing the size can be achieved by changing only one value. This is pretty basic good programming practice.

The use of constants in call to size() seem technically feasible, because it could be solved in the preprocessing stage.

Thank you very much in advance!

**Comment by GoToLoop** [comment]
http://forum.Processing.org/two/discussion/12105/ok-to-use-constants-in-call-to-size-in-immediate-mode

**Comment by benfry** [comment]
The practice is debatable—you'd be better of teaching students to respond to the current state of the `width` and `height` variables, given the wide range of devices for which they'll eventually be developing. This is why `width` and `height` exist. 

But for your specific use case, you can use `settings()` instead: https://processing.org/reference/settings_.html

**Comment by jarmohurri** [comment]
Greetings Ben.

Maybe my case is a bit special, but let me explain why neither settings() nor width and height work as well as constants. If for nothing else, then because this might be of general interest to you.

We are going to teach programming to children of ages 13-16 using Processing as their second programming environment. At the minimum this will concern hundreds of children; we might end up teaching nationally, which would then mean tens of thousands annually.

When we teach the use of constants to them (by varying parameters such as width and height in sketches) they will know nothing about defining functions or global variables. Hence settings() would confuse them - because that would require defining a function. Use of width and height would also confuse them: you put values into a function - size() - and suddenly they are available in these automatically named variables.

Defining constants clearly in the program, and using them all the way in the code would be conceptually *by far* the best approach. But if Processing will not allow for this to happen, we will just have to vary other properties instead of width and height. Width and height would have just been a very intuitive starting point.

All the best,

Jarmo Hurri
lecturer, PhD

**Comment by GoToLoop** [comment]
You've forgotten to mention that you're interested solely in the "Immediate Mode" here too.
Indeed **size()** should accept variables & expressions when running w/o defining functions even in P3.

**Comment by benfry** [comment]
@jarmohurri It's not for lack of wanting to support that use case, it's simply that it makes things exceptionally complicated and messy. Some of it is explained [here](https://github.com/processing/processing/wiki/Changes-in-3.0#things-that-may-break-your-2x-sketches) and [here](https://github.com/processing/processing/blob/master/core/README.md).

Because it's possible to draw inside `setup()`, we need to have the rendering context (Java2D, OpenGL, etc) already set up and specified. In order to make sure they're specified already, we move `size()` into the `settings()` method behind the scenes. In earlier releases, We called `setup()` and then threw an exception and restarted the sketch in order to change the renderer. This hack was untenable for long-term stability, and caused fits for other environments that have been developed (Python, Ruby, etc). 

Bottom line, I just wouldn't recommend this particular use case, but it's not implemented this way to spite anyone, or for philosophical differences, or for lack of enormous amounts of work trying to support it.

**Comment by jarmohurri** [comment]
GoToLoop: thanks for the clarification. I have been writing to two different forums today, and forgot to mention immediate mode here. (We want our code to be supported in the long run, so we will go for version 3 if at all possible.)

Ben: I understand. Thank you for delivering such a great piece of software for us and - as I am very happy to notice - continuing to support it and develop it further. If our teaching goes nationwide I will let you know.

All the best,

Jarmo
